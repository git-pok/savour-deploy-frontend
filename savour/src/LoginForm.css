.LoginForm-bg-img {
  width: 100vw;
  height: 100vh;
  margin: 0px 0px;
}

.LoginForm {
  display: flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;
  width: 50%;
  margin: 100px auto 0px auto;
  padding: 50px 0px;
  background-color: rgb(78, 42, 42);
  border-radius: 10px;
}

.LoginForm-field {
  display: block;
}
  
.LoginForm-h1 {
  text-align: center;
  display: block;
  font-size: 60px;
  margin: 0px 0px;
  padding-top: 120px;
}

.LoginForm-field label {
  display: block;
  width: 260px;
  height: 30px;
  margin-right: 30px;
  color: rgb(255, 255, 255);
}
  
.LoginForm-field input {
  width: 260px;
  height: 30px;
  margin-bottom: 30px;
}

.LoginForm-submit {
  display: block;
  width: 100%;
  margin-top: 10px;
}

.LoginForm-submit button {
  display: block;
  width: 260px;
  height: 30px;
  margin: 0px auto;
  margin-top: 10px;
  border: 3px inset rgb(223, 216, 85);
  background-color: rgb(223, 216, 85);
  color: rgb(0, 0, 0);
  transition: .5s ease-in-out background-color;
}
  
.LoginForm button:hover {
  border-color: rgb(255, 255, 255);
  background-color: rgb(255, 255, 255);
  color: rgb(0, 0, 0);
}