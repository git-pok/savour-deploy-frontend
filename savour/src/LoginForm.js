import { useState, useEffect, useContext } from 'react';
import { Redirect, useHistory } from 'react-router-dom';
import Message from './Message.js';
import SavourApi from './models/SavourApi.js';
import useLocalStorage from './hooks/useLocalStorage.js';
import useToggleState from './hooks/useToggleState.js';
import jwt_decode from 'jwt-decode';
import UserContext from './context/UserContext.js';
import './LoginForm.css';

/**
 * LoginForm
 * Props: formValsCheck
 *    formValsCheck: function to check if
 *     form props are missing.
 * Renders login form.
*/
const LoginForm = ({ formValsCheck }) => {
  const initialState = {
      username: "", password: ""
  };

  const [ formErrMsg, setFormErrMsg ] = useState(null);
  const [ formData, setFormData ] = useState(initialState);
  const [ isSubmitted, setIsSubmitted ] = useToggleState(false);
  const [ invalidForm, setInvalidForm ] = useToggleState(false);
  const { usrData, setUsrData } = useContext(UserContext);
  const history = useHistory();

  useEffect(() => {
    const login = async () => {
      try {
        const { username, password } = formData;
        const data = { username, password };
        const regResult = await SavourApi.request("post", "/users/login", data);
        const token = regResult.data[0].user.token;
        SavourApi.token = token;
        const payload = await jwt_decode(token);
        payload.token = token;

        setUsrData(data => (
          payload
        ));
        // Set isSubmitted to false.
        setIsSubmitted();
        setFormData(() => initialState);
        history.push("/");
      } catch (err) {
        console.log("ERROR", err);
        // Define variable for API error.
        const error = err.response.data ? err.response.data.error.message : null;
        const isErrObj = typeof err.message === "object";
        const isErrorObj = typeof error === "object";
        const errMsg = isErrObj ? null : err.message;
        const errorMsg = isErrorObj ? null : error;
        setFormErrMsg(() => errorMsg || errMsg || "Error!");
        setInvalidForm();
        setTimeout(setInvalidForm, 3000);
        // Set isSubmitted to false.
        setTimeout(setIsSubmitted, 3000);
      }
    }

    if (isSubmitted) login();

  }, [isSubmitted])

  const handleChange = (evt) => {
    const { name, value } = evt.target;

    setFormData(data => ({
      ...data,
      [name]: value
    }))
  }

  const handleSubmit = async (evt) => {
    evt.preventDefault();
    // Checks if form props are missing.
    const isValMsn = formValsCheck(["username", "password"], formData);
    // If props are missing set invalidForm.
    if (isValMsn) {
      setFormErrMsg(() => "All fields must be complete!");
      setInvalidForm();
      setTimeout(setInvalidForm, 3000);
    }
    // Set isSubmitted to true if all props exist.
    else setIsSubmitted();
  }

  return (
    <>
    <h1 className="LoginForm-h1">Login</h1>
    <form onSubmit={handleSubmit} className="LoginForm">
      <div className="LoginForm-field">
        <label htmlFor="username">Username</label>
        <input
            type="text"
            id="username"
            onChange={handleChange}
            value={formData.username}
            name="username"
            placeholder="Type a username"
            autoComplete="username"></input>
      </div>
      <div className="LoginForm-field">
        <label htmlFor="password">Password</label>
        <input
            type="password"
            id="password"
            onChange={handleChange}
            value={formData.password}
            name="password"
            placeholder="Type a password"
            autoComplete="current-password"></input>
      </div>
      <div className="LoginForm-submit">
      {
        invalidForm &&
        <Message msgObj={
          {
            class: "fail",
            msg: formErrMsg ? formErrMsg : "Error!"
          }
        } />
      }
        <button data-testid="sbmtBtn">SUBMIT</button>
      </div>
    </form>
    </>
  );
}

export default LoginForm;