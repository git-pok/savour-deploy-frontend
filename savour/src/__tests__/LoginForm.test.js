import LoginForm from '../LoginForm.js';
import { render, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import UserContext from '../context/UserContext.js';


const setUsrData = () => {
  console.log("Test!");
};


test("renders LoginForm without crashing", () => {
  render(
    <MemoryRouter>
      <UserContext.Provider value={{setUsrData}}>
        <LoginForm />
      </UserContext.Provider>
    </MemoryRouter>
  );
});

test("renders snapshot without crashing", () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserContext.Provider value={{setUsrData}}>
        <LoginForm />
      </UserContext.Provider>
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});

test("login form elements exist", () => {
  const { queryByLabelText } = render(
    <MemoryRouter intitialEntries={["/login"]}>
      <UserContext.Provider value={{setUsrData}}>
        <LoginForm />
      </UserContext.Provider>
    </MemoryRouter>
  );

  const signInUsrname = queryByLabelText("Username");
  const signInPassword = queryByLabelText("Password");
  expect(signInUsrname).toBeInTheDocument();
  expect(signInPassword).toBeInTheDocument();
});