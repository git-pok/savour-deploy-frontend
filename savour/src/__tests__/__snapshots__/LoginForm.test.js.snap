// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot without crashing 1`] = `
<DocumentFragment>
  <h1
    class="LoginForm-h1"
  >
    Login
  </h1>
  <form
    class="LoginForm"
  >
    <div
      class="LoginForm-field"
    >
      <label
        for="username"
      >
        Username
      </label>
      <input
        autocomplete="username"
        id="username"
        name="username"
        placeholder="Type a username"
        type="text"
        value=""
      />
    </div>
    <div
      class="LoginForm-field"
    >
      <label
        for="password"
      >
        Password
      </label>
      <input
        autocomplete="current-password"
        id="password"
        name="password"
        placeholder="Type a password"
        type="password"
        value=""
      />
    </div>
    <div
      class="LoginForm-submit"
    >
      <button
        data-testid="sbmtBtn"
      >
        SUBMIT
      </button>
    </div>
  </form>
</DocumentFragment>
`;
